### DISCUSSION TITLE
?index added on submit for json

### DISCUSSION THREAD

**Discussion by daniele-zurico** [remix-run/remix#2588] [Q&A] [answered]
Hello everyone,
I'm mainly following the example provided in the documentation but I found something really weird.
If inside my `/routes` I create and `index.tsx` with the following content:

```tsx
import { Form, json, useActionData } from 'remix';

export const action = () => {
  return json({ error: 'Error' });
};

export default function Index() {
  const data = useActionData();

  return (
    <div style={{ fontFamily: 'system-ui, sans-serif', lineHeight: '1.4' }}>
      <h1>Welcome to Remix</h1>
      <Form method="post">
        <h2>Submit to the same URL</h2>
        <button>Submit</button>
      </Form>
      <pre>
        <code>{JSON.stringify(data, null, 2)}</code>
      </pre>
    </div>
  );
}
```
When I submit the form by default I see in the url that `http://baseUrl/?index` is added.
How I can avoid that `?index` is added?

**Comment by ryanflorence** [comment, marked as answer]
`root.jsx` and `routes/index.jsx` share the same URL. We use `?index` to disambiguate which one you're trying to post to.

The default `action` of a `<Form>` is *the route it's rendered in*. So in your case, it's defaulting to `<Form action="?index" />` so that it posts to the index route's action where you rendered it.

So you can either:

- Point the form to the root route `<Form action="/">` (or any other route `<Form action="/other" />`
- Move your Form and action into the `root.tsx` route.

**Reply by daniele-zurico** [reply]
Hi @ryanflorence thanks for the reply... yes I realized that changing from index.tsx to actual-page.tsx will fix my problem... however I was trying to follow the documentation that suggests:
```
/path-to-page
  ->index.tsx
```
(of course, this is simplifying my code... in reality, I've nested routes)
if I use this pattern I am pretty much not able to achieve what you suggest because I need to redirect to a different page. What I want instead is to stay on the same url and display the error message.
```tsx
<code>{JSON.stringify(data, null, 2)}</code>
```

In nutshell:
1. if we use the index.tsx under the folder you'll always deal with the `?index`
2. if we use the page.tsx it will fix the problem but all the times you need to add a sub-route you need to refactor your code changing the page.tsx in a folder and add a sub-route.

(I'm not sure but from my point of view is a technical limitation of the framework)

This is a quick repro: https://codesandbox.io/s/keen-banach-ou2ox5

**Reply by ryanflorence** [reply]
Move the action to the parent route, change the form action to the parent route.

```jsx
// routes/index.jsx
<Form action="/" />
```

```jsx
// root.jsx
export function action() {}
```

**Reply by daniele-zurico** [reply]
Thanks for the reply @ryanflorence, I'm probably doing something wrong but still doesn't work as expected.
I updated my repro:  https://codesandbox.io/s/goofy-rumple-ejgssi?file=/app/root.jsx

but despite the form redirect to the same page I'm not able to see:
```tsx
<code>{JSON.stringify(data, null, 2)}</code>
```
inside my `routes/my-page/index.tsx`

**Reply by houmark** [reply]
Anyone (like me) hitting this question from Google who wish to get rid of ?index in the URL. Consider using the `useFetcher()` hook. Docs [here](https://remix.run/docs/en/v1/api/remix#usefetcher).

**Reply by Barlow1** [reply]
> Anyone (like me) hitting this question from Google who wish to get rid of ?index in the URL. Consider using the `useFetcher()` hook. Docs [here](https://remix.run/docs/en/v1/api/remix#usefetcher).

Yo @houmark, what are the odds I run into this issue, find your comment from 5 hrs ago and swap out Form with useFetcher().Form and fix my issue in minutes? Thank you my friend 🙏🏻

My particular instance was occurring in a nested folder with an index.tsx file so Ryans advice didn't totally apply to me, but this worked for me.
